File: android/Streamo/app/src/main/java/com/streamo/app/util/Release.kt
package com.streamo.app.util

import com.streamo.app.data.remote.dto.TmdbItem
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

object Release {
    private val calendar: Calendar get() = Calendar.getInstance()

    private val longFmt = SimpleDateFormat("d MMMM yyyy", Locale.ITALIAN)
    private val shortFmt = SimpleDateFormat("d MMM", Locale.ITALIAN)

    fun parseDate(s: String?): Date? {
        if (s.isNullOrEmpty() || s.length < 10) return null
        val p = s.substring(0, 10).split("-")
        if (p.size != 3) return null
        val y = p[0].toIntOrNull() ?: return null
        val m = p[1].toIntOrNull() ?: return null
        val d = p[2].toIntOrNull() ?: return null
        return calendar.apply { set(y, m - 1, d, 0, 0, 0); set(Calendar.MILLISECOND, 0) }.time
    }

    fun isFuture(d: Date): Boolean {
        val today = calendar.apply { set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0) }.time
        return d > today
    }

    fun longDate(d: Date): String = longFmt.format(d)
    fun shortDate(d: Date): String = shortFmt.format(d)

    fun watchlistMeta(item: TmdbItem, type: String): Pair<Boolean, String?> {
        val titleDate = if (type == "movie") item.releaseDate else item.firstAirDate
        val d = parseDate(titleDate)
        if (d != null && isFuture(d)) {
            return Pair(true, if (type == "movie") "Esce il ${longDate(d)}" else "Dal ${longDate(d)}")
        }
        if (type == "tv") {
            val nd = parseDate(item.nextEpisodeToAir?.airDate)
            if (nd != null && isFuture(nd)) {
                return Pair(false, "Nuovo ep. ${shortDate(nd)}")
            }
        }
        return Pair(false, null)
    }

    fun isUpcoming(item: TmdbItem, type: String): Boolean = watchlistMeta(item, type).first

    fun upcomingBadge(item: TmdbItem, type: String): String? {
        if (!isUpcoming(item, type)) return null
        return if (type == "movie") "Prossimamente" else "Nuova serie"
    }

    fun compactStatus(item: TmdbItem, type: String): String? {
        val meta = watchlistMeta(item, type)
        if (meta.second != null) return meta.second
        if (type != "tv") return null
        val next = findNextSeason(item) ?: return null
        return "Stagione ${next.first} il ${shortDate(next.second)}"
    }

    fun fullStatus(item: TmdbItem, type: String): String {
        if (type == "tv") {
            val d = parseDate(item.firstAirDate)
            if (d != null && isFuture(d)) return "Nuova serie dal ${longDate(d)}."
        }
        if (type == "movie") {
            val d = parseDate(item.releaseDate)
            if (d != null && isFuture(d)) return "Esce il ${longDate(d)}."
            return ""
        }
        val nea = item.nextEpisodeToAir
        val nd = parseDate(nea?.airDate)
        if (nd != null && isFuture(nd)) {
            val s = nea?.seasonNumber?.toString() ?: "?"
            val e = nea?.episodeNumber?.toString() ?: "?"
            return "Prossimo episodio: S$s E$e in uscita il ${longDate(nd)}."
        }
        val next = findNextSeason(item)
        if (next != null) {
            return "Prossima stagione: Stagione ${next.first} in uscita il ${longDate(next.second)}."
        }
        if (item.status == "Ended" || item.status == "Canceled") return "Serie conclusa."
        return ""
    }

    private fun findNextSeason(item: TmdbItem): Pair<Int, Date>? {
        val lastSeason = TVLogic.effectiveLastEpisode(item)?.first ?: 0
        return item.seasons.orEmpty()
            .filter { it.seasonNumber > 0 }
            .mapNotNull { s ->
                val d = parseDate(s.airDate)
                if (d != null && s.seasonNumber > lastSeason) Pair(s.seasonNumber, d) else null
            }
            .sortedBy { it.second }
            .firstOrNull { isFuture(it.second) }
    }
}

File: ios/Streamo/Streamo/Domain/Release.swift
import Foundation

/// Release / upcoming copy — port of the web `media-release.util.ts` +
/// `release-format.ts` (getWatchlistReleaseMeta). Produces the Italian
/// "Esce il …", "Nuovo ep. …", "Prossima stagione …" strings used on cards
/// (`nextReleaseText`) and the detail page release line.
enum Release {
    // MARK: Date helpers

    private static var calendar: Calendar {
        var c = Calendar(identifier: .gregorian); c.timeZone = .current; return c
    }

    static func parseDate(_ s: String?) -> Date? {
        guard let s, s.count >= 10 else { return nil }
        let p = s.prefix(10).split(separator: "-")
        guard p.count == 3, let y = Int(p[0]), let m = Int(p[1]), let d = Int(p[2]) else { return nil }
        return calendar.date(from: DateComponents(year: y, month: m, day: d))
    }

    static func isFuture(_ d: Date) -> Bool {
        calendar.startOfDay(for: d) > calendar.startOfDay(for: Date())
    }

    private static let longFmt: DateFormatter = {
        let f = DateFormatter(); f.locale = Locale(identifier: "it_IT"); f.setLocalizedDateFormatFromTemplate("d MMMM yyyy"); return f
    }()
    private static let shortFmt: DateFormatter = {
        let f = DateFormatter(); f.locale = Locale(identifier: "it_IT"); f.setLocalizedDateFormatFromTemplate("d MMM"); return f
    }()

    static func longDate(_ d: Date) -> String { longFmt.string(from: d) }
    static func shortDate(_ d: Date) -> String { shortFmt.string(from: d) }

    // MARK: Public API

    /// Port of getWatchlistReleaseMeta.
    static func watchlistMeta(_ item: TmdbItem, _ type: MediaType) -> (isUpcoming: Bool, text: String?) {
        let titleDate = type == .movie ? item.releaseDate : item.firstAirDate
        if let d = parseDate(titleDate), isFuture(d) {
            return (true, type == .movie ? "Esce il \(longDate(d))" : "Dal \(longDate(d))")
        }
        if type == .tv, let d = parseDate(item.nextEpisodeToAir?.airDate), isFuture(d) {
            return (false, "Nuovo ep. \(shortDate(d))")
        }
        return (false, nil)
    }

    static func isUpcoming(_ item: TmdbItem, _ type: MediaType) -> Bool {
        watchlistMeta(item, type).isUpcoming
    }


    /// Compact line for cards (`nextReleaseText`). Port of getCompactReleaseStatusText.
    static func compactStatus(_ item: TmdbItem, _ type: MediaType) -> String? {
        if let t = watchlistMeta(item, type).text { return t }
        guard type == .tv, let next = findNextSeason(item) else { return nil }
        return "Stagione \(next.season) il \(shortDate(next.date))"
    }

    /// Full sentence for the detail page. Port of getFullReleaseStatusText.
    static func fullStatus(_ item: TmdbItem, _ type: MediaType) -> String {
        if type == .tv, let d = parseDate(item.firstAirDate), isFuture(d) {
            return "Nuova serie dal \(longDate(d))."
        }
        if type == .movie {
            if let d = parseDate(item.releaseDate), isFuture(d) { return "Esce il \(longDate(d))." }
            return ""
        }
        if let nea = item.nextEpisodeToAir, let d = parseDate(nea.airDate), isFuture(d) {
            let s = nea.seasonNumber.map(String.init) ?? "?"
            let e = nea.episodeNumber.map(String.init) ?? "?"
            return "Prossimo episodio: S\(s) E\(e) in uscita il \(longDate(d))."
        }
        if let next = findNextSeason(item) {
            return "Prossima stagione: Stagione \(next.season) in uscita il \(longDate(next.date))."
        }
        if item.status == "Ended" || item.status == "Canceled" { return "Serie conclusa." }
        return ""
    }

    private static func findNextSeason(_ item: TmdbItem) -> (season: Int, date: Date)? {
        let lastSeason = TVLogic.effectiveLastEpisode(item)?.season ?? 0
        return (item.seasons ?? [])
            .filter { $0.seasonNumber > 0 }
            .compactMap { s -> (season: Int, date: Date)? in
                guard let d = parseDate(s.airDate), s.seasonNumber > lastSeason else { return nil }
                return (s.seasonNumber, d)
            }
            .sorted { $0.date < $1.date }
            .first { isFuture($0.date) }
    }
}
